### PR TITLE
MVP: Handle extra fields to export

### LINKED ISSUE #277: Handle new additional fields exports
## Related issues

https://github.com/kobotoolbox/kpi/issues/3380

### PR DESCRIPTION
**Pull request by joshuaberetta**
Tasks:
- [x] export basic form with analysis questions and responses
- [x] group survey questions with their associated analysis questions in export
- [x] allow for analysis questions to be toggled on or off for export
- [x] extend test fixtures for analysis forms
- [x] fall back on analysis form question name if translated label not available
- [x] test:
  - [x] works with multiple select in main survey and different types of `multiple_select` (both, details, summary)
  - [x] analysis form can have select type questions even if the original survey has no select questions
  - [x] formpack internals
  - [x] normal groups
  - [x] form changes across versions
  - [x] auto report ignores analysis questions
  - [x] translations
  - [x] export without analysis fields

## Description

Allow for translations, transcriptions and coding questions to be integrated into the export.

## Related issues

closes #277

### CODE REVIEW

File: src/formpack/pack.py [hunk 1]
```diff
@@ -176,6 +179,9 @@ def load_version(self, schema):
 
         self.versions[form_version.id] = form_version
 
+    def extend_survey(self, analysis_form: Dict) -> None:
```
Comment (noliveleger) [inline src/formpack/pack.py:182]: I used to use `: dict` . I never thought about using `typing import Dict` .

File: src/formpack/schema/fields.py [hunk 2]
```diff
@@ -45,7 +64,7 @@ def __init__(
         if has_stats is not None:
             self.has_stats = has_stats
         else:
-            self.has_stats = data_type != 'note'
+            self.has_stats = data_type != "note" and not self.analysis_question
```
Comment (noliveleger) [inline src/formpack/schema/fields.py:67]: Why changing `'` to `"`?

File: src/formpack/schema/fields.py [hunk 3]
```diff
@@ -139,7 +158,7 @@ def _get_label(
         # even if `lang` can be None, we don't want the `label` to be None.
         label = self.labels.get(lang, self.name)
         # If `label` is None, no matches are found, so return `field` name.
-        return self.name if label is None else label
+        return label or self.name
```
Comment (noliveleger) [inline src/formpack/schema/fields.py:161]: 👀 , even If I prefer your change, it does not do the same thing. if `label` is empty (not None), `self.name` would be returned. Is it really what we want? I guess so.

File: src/formpack/schema/fields.py [hunk 4]
```diff
@@ -424,6 +458,72 @@ def get_substats(
...
 class TextField(ExtendedFormField):
+    @property
+    def _is_transcript(self):
+        return getattr(self, 'analysis_type', '') == ANALYSIS_TYPE_TRANSCRIPT
+
+    @property
+    def _is_translation(self):
+        return getattr(self, 'analysis_type', '') == ANALYSIS_TYPE_TRANSLATION
+
+    def get_labels(
```
Comment (noliveleger) [inline src/formpack/schema/fields.py:469]: Did we say we would put "private/protected" method after public ones? 
We should keep the methods sorted alphabetically too. This file starts to be really big, It's easier to find method inside classes.

File: src/formpack/version.py [hunk 5]
```diff
@@ -42,7 +51,33 @@ def get(self, key, default=None):
         return self._vals.get(key, default)
 
 
-class FormVersion:
+class BaseForm:
+    @staticmethod
+    def _get_translations(content: Dict[str, List]) -> List[str]:
```
Comment (noliveleger) [inline src/formpack/version.py:56]: 👀  alphabetically sorted?

File: src/formpack/version.py [hunk 6]
```diff
@@ -345,3 +368,83 @@ def to_xml(self, warnings=None):
         )
 
         return survey._to_pretty_xml()  # .encode('utf-8')
+
+
+class AnalysisForm(BaseForm):
```
Comment (noliveleger) [inline src/formpack/version.py:373]: Let's move it before `FormVersion` :-)

File: src/formpack/pack.py (hunk 1 above)
Comment (joshuaberetta) [reply src/formpack/pack.py:182]: Interesting... which do you prefer? With `Dict` we can also be more specific like `Dict[str, str]`

File: src/formpack/schema/fields.py (hunk 3 above)
Comment (joshuaberetta) [reply src/formpack/schema/fields.py:161]: The change was intentional as I don't think the previous logic was correct. AFAIR I was causing some of my new tests to fail so I changed it 👌

File: src/formpack/schema/fields.py (hunk 4 above)
Comment (joshuaberetta) [reply src/formpack/schema/fields.py:469]: Good point 👍

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: src/formpack/reporting/export.py [hunk 7]
```diff
@@ -90,6 +90,17 @@ def __init__(
...
 
+        self.t_lang_codes = []
+        _filter_fields = []
+        for item in self.filter_fields:
+            item = re.sub(r'^_supplementalDetails/', '', item)
+            if item.split('/')[-1].startswith('transcript'):
+                self.t_lang_codes.append(item.split('_')[-1])
+                _filter_fields.append(f"{item.split('/')[0]}/transcript")
+            else:
+                _filter_fields.append(item)
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply src/formpack/reporting/export.py:101]: @joshuaberetta what about this: 
```python
            item = re.sub(r'^_supplementalDetails/', '', item)
            # use explicit names for variable below. 
            # `_` is dummy content, we don't use.
            my_variable, *_, last_variable = item.split('/')
            *_, my_other_variable  = item.split('_')
            
            if last_variable.startswith('transcript'):
                self.t_lang_codes.append(my_other_variable)
                _filter_fields.append(f"{my_variable}/transcript")
            else:
                _filter_fields.append(item)
```

File: src/formpack/pack.py (hunk 1 above)
Comment (noliveleger) [reply src/formpack/pack.py:182]: We are now imported `from __future__ import annotations` 

`def extend_survey(self, analysis_form: dict) -> None:`

and `dict[str, str]` should work too.

File: src/formpack/schema/fields.py (hunk 3 above)
Comment (jnm) [reply src/formpack/schema/fields.py:161]: I somewhat doubt that it's related, but we need careful handling of labels and more unit test coverage. I've found a regression related to choice names with missing translations: https://chat.kobotoolbox.org/#narrow/stream/4-Kobo-Dev/topic/formpack.20and.20missing.20labels/near/173165. I'll make a correction or open a new issue as needed.

File: src/formpack/reporting/export.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jnm) [reply src/formpack/reporting/export.py:101]: now tracked by #306

File: src/formpack/pack.py (hunk 1 above)
Comment (jnm) [reply src/formpack/pack.py:182]: now tracked by #306

File: src/formpack/schema/fields.py (hunk 3 above)
Comment (jnm) [reply src/formpack/schema/fields.py:161]: #307

**Pull request merged**